jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannelState.java

```diff
@@ -1398,11 +1367,11 @@ public void onReadUnready()
             switch (_inputState)
             {
                 case IDLE:
-                case UNREADY:
-                case READY:  // READY->UNREADY is needed by AsyncServletIOTest.testStolenAsyncRead
                     _inputState = InputState.UNREADY;
                     break;
 
+                case READY:
```

Comment:
Just thinking out loud here... and maybe this should be in a comment.  We only call this method if:
 + some thread has called isReady
 + We were not already UNREADY
 + An attempt to produce a chunk returned null
 + We are going to demand immediately when this method returns.

So if we are READY, can we ever validly end up here?    We enter READY only when a readListener is first registered or when the demand `Runnable` is called.    The first case could have nothing to read, and the later case might have a chunk that is either empty.   We are meant to leave READY if somebody has read a non null chunk, but they often do that after a superfluous call to isReady.  So I'm not convinced we can't validly be here in READY state.... specially if some other read steals our chunk.

Reply:
Before `unhandle()` returns `Action.READ_CALLBACK`, it flips the inputState back to `IDLE` so it shouldn't be possible to transition from `READY` to `UNREADY`. This is now documented.